File: backend/api/model/workbench/tool/tool.go
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tool

type MCPToolDefinition struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	InputSchema string `json:"input_schema"`
}

type MCPServerSourceType string

const (
	MCPServerSourceTypeCustom   MCPServerSourceType = "custom"
	MCPServerSourceTypeOfficial MCPServerSourceType = "official"
)

func (s MCPServerSourceType) Valid() bool {
	return s == MCPServerSourceTypeCustom || s == MCPServerSourceTypeOfficial
}

type MCPResource struct {
	// URI is an internal runtime locator. It must never be serialized through a
	// Workbench response or export.
	URI         string `json:"-"`
	ResourceID  string `json:"resource_id,omitempty"`
	Name        string `json:"name"`
	Description string `json:"description"`
	MIMEType    string `json:"mime_type"`
}

type MCPPromptArgument struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	Required    bool   `json:"required"`
}

type MCPPrompt struct {
	Name        string               `json:"name"`
	Description string               `json:"description"`
	Arguments   []*MCPPromptArgument `json:"arguments"`
}

type MCPToolServer struct {
	ServerID        int64                `json:"server_id,string"`
	SpaceID         int64                `json:"space_id,string"`
	CreatorID       int64                `json:"creator_id,string"`
	SourceType      MCPServerSourceType  `json:"source_type"`
	Name            string               `json:"name"`
	Description     string               `json:"description"`
	ServerType      string               `json:"server_type"`
	Enabled         bool                 `json:"enabled"`
	Config          string               `json:"config"`
	Auth            string               `json:"auth"`
	Tools           []*MCPToolDefinition `json:"tools"`
	Resources       []*MCPResource       `json:"resources"`
	Prompts         []*MCPPrompt         `json:"prompts"`
	HealthStatus    string               `json:"health_status"`
	HealthCheckedAt int64                `json:"health_checked_at"`
	HealthLatencyMs int64                `json:"health_latency_ms"`
	HealthError     string               `json:"health_error"`
	CreatedAt       int64                `json:"created_at"`
	UpdatedAt       int64                `json:"updated_at"`
}

type UpsertMCPToolServerRequest struct {
	ServerID    int64                `path:"server_id" json:"server_id,string,omitempty"`
	SpaceID     int64                `json:"space_id,string,required" query:"space_id"`
	Name        string               `json:"name,required"`
	Description string               `json:"description,omitempty"`
	ServerType  string               `json:"server_type,required"`
	Enabled     bool                 `json:"enabled"`
	Config      string               `json:"config,required"`
	Auth        string               `json:"auth,omitempty"`
	Tools       []*MCPToolDefinition `json:"tools,required"`
}

type ListMCPToolServersRequest struct {
	SpaceID int64 `query:"space_id,required"`
}

type ListMCPToolRegistryEntriesRequest struct {
	SpaceID int64 `query:"space_id,required"`
}

type GetMCPToolServerRequest struct {
	ServerID int64 `path:"server_id,required"`
}

type TestMCPToolCallRequest struct {
	ServerID  int64  `path:"server_id,required" json:"-"`
	ToolName  string `json:"tool_name,required"`
	Arguments string `json:"arguments,required"`
}

type DiscoverMCPToolServerRequest struct {
	ServerID int64 `path:"server_id,required" json:"-"`
}

type ExportMCPToolServerRequest struct {
	ServerID int64 `path:"server_id,required" json:"-"`
}

type ListMCPRuntimeAuditEventsRequest struct {
	ServerID int64  `path:"server_id,required" json:"-"`
	Limit    int32  `query:"limit"`
	Cursor   string `query:"cursor"`
}

type ListMCPToolServersData struct {
	Servers   []*MCPToolServer `json:"servers"`
	Total     int64            `json:"total"`
	CanManage bool             `json:"can_manage"`
}

type MCPToolRegistryEntry struct {
	Name            string `json:"name"`
	Source          string `json:"source"`
	Category        string `json:"category"`
	Visibility      string `json:"visibility"`
	ServerID        int64  `json:"server_id,string"`
	ServerName      string `json:"server_name"`
	ToolName        string `json:"tool_name"`
	Description     string `json:"description"`
	InputSchema     string `json:"input_schema"`
	Enabled         bool   `json:"enabled"`
	HealthStatus    string `json:"health_status"`
	HealthCheckedAt int64  `json:"health_checked_at"`
	HealthLatencyMs int64  `json:"health_latency_ms"`
	HealthError     string `json:"health_error"`
}

type ListMCPToolRegistryEntriesData struct {
	Tools []*MCPToolRegistryEntry `json:"tools"`
	Total int64                   `json:"total"`
}

type TestMCPToolCallData struct {
	Status    string `json:"status"`
	Output    string `json:"output"`
	LatencyMs int64  `json:"latency_ms"`
}

type DiscoverMCPToolServerData struct {
	Tools     []*MCPToolDefinition `json:"tools"`
	Resources []*MCPResource       `json:"resources"`
	Prompts   []*MCPPrompt         `json:"prompts"`
}

// ExportMCPToolServerData intentionally omits authentication material. It is a
// portable control-plane representation, not a dump of the persisted record.
type ExportMCPToolServerData struct {
	Name        string               `json:"name"`
	Description string               `json:"description"`
	ServerType  string               `json:"server_type"`
	Config      string               `json:"config"`
	Tools       []*MCPToolDefinition `json:"tools"`
	Resources   []*MCPResource       `json:"resources"`
	Prompts     []*MCPPrompt         `json:"prompts"`
}

type MCPRuntimeAuditEvent struct {
	EventID      string `json:"event_id"`
	ActorID      int64  `json:"actor_id,string"`
	ToolName     string `json:"tool_name"`
	Status       string `json:"status"`
	LatencyMs    int64  `json:"latency_ms"`
	ErrorCode    string `json:"error_code,omitempty"`
	ErrorSummary string `json:"error_summary,omitempty"`
	CreatedAt    int64  `json:"created_at"`
	CompletedAt  int64  `json:"completed_at,omitempty"`
}

type ListMCPRuntimeAuditEventsData struct {
	Events     []*MCPRuntimeAuditEvent `json:"events"`
	NextCursor string                  `json:"next_cursor"`
}

type MCPToolServerResponse struct {
	Data *MCPToolServer `json:"data,omitempty"`
	Code int64          `json:"code"`
	Msg  string         `json:"msg"`
}

type ListMCPToolServersResponse struct {
	Data *ListMCPToolServersData `json:"data,omitempty"`
	Code int64                   `json:"code"`
	Msg  string                  `json:"msg"`
}

type ListMCPToolRegistryEntriesResponse struct {
	Data *ListMCPToolRegistryEntriesData `json:"data,omitempty"`
	Code int64                           `json:"code"`
	Msg  string                          `json:"msg"`
}

type TestMCPToolCallResponse struct {
	Data *TestMCPToolCallData `json:"data,omitempty"`
	Code int64                `json:"code"`
	Msg  string               `json:"msg"`
}

type DiscoverMCPToolServerResponse struct {
	Data *DiscoverMCPToolServerData `json:"data,omitempty"`
	Code int64                      `json:"code"`
	Msg  string                     `json:"msg"`
}

type ExportMCPToolServerResponse struct {
	Data *ExportMCPToolServerData `json:"data,omitempty"`
	Code int64                    `json:"code"`
	Msg  string                   `json:"msg"`
}

type ListMCPRuntimeAuditEventsResponse struct {
	Data *ListMCPRuntimeAuditEventsData `json:"data,omitempty"`
	Code int64                          `json:"code"`
	Msg  string                         `json:"msg"`
}

File: frontend/packages/arch/api-schema/src/idl/workbench/tool.ts
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createAPI } from './../../api/config';

export interface MCPToolDefinition {
  name: string;
  description: string;
  input_schema: string;
}

export type MCPServerSourceType = 'custom' | 'official';

export interface MCPResource {
	resource_id: string;
	name: string;
  description: string;
  mime_type: string;
}

export interface MCPPromptArgument {
  name: string;
  description: string;
  required: boolean;
}

export interface MCPPrompt {
  name: string;
  description: string;
  arguments: MCPPromptArgument[];
}

export interface MCPToolServer {
  server_id: string;
  space_id: string;
  creator_id: string;
  source_type: MCPServerSourceType;
  name: string;
  description: string;
  server_type: string;
  enabled: boolean;
  config: string;
  auth: string;
  tools: MCPToolDefinition[];
  resources: MCPResource[];
  prompts: MCPPrompt[];
  health_status: string;
  health_checked_at: number;
  health_latency_ms: number;
  health_error: string;
  created_at: number;
  updated_at: number;
}

export interface UpsertMCPToolServerRequest {
  server_id?: string;
  space_id: string;
  name: string;
  description?: string;
  server_type: string;
  enabled: boolean;
  config: string;
  auth?: string;
  tools: MCPToolDefinition[];
}

export interface ListMCPToolServersRequest {
  space_id: string;
}

export type MCPOfficialInstallStatus =
  | 'available'
  | 'installed'
  | 'needs_migration';

export type MCPOfficialAvailability = 'installable' | 'adapter_required';

export interface MCPOfficialCredentialField {
  key: string;
  label: string;
  description: string;
  placeholder: string;
  required: boolean;
  secret: boolean;
}

export interface MCPOfficialInstallation {
  server_id: string;
  enabled: boolean;
  health_status: string;
  health_checked_at: number;
  health_latency_ms: number;
  updated_at: number;
}

export interface MCPOfficialCatalogEntry {
  catalog_id: string;
  name: string;
  description: string;
  icon_url: string;
  publisher: string;
  source: string;
  server_type: string;
  tools: MCPToolDefinition[];
  credential_fields: MCPOfficialCredentialField[];
  availability: MCPOfficialAvailability;
  availability_reason: string;
  install_status: MCPOfficialInstallStatus;
  installation?: MCPOfficialInstallation;
}

export interface ListMCPOfficialCatalogRequest {
  space_id: string;
}

export interface InstallMCPOfficialCatalogRequest {
  catalog_id: string;
  space_id: string;
  credentials: Record<string, string>;
}

export interface ListMCPOfficialCatalogData {
  entries: MCPOfficialCatalogEntry[];
  total: number;
  can_manage: boolean;
}

export interface ListMCPOfficialCatalogResponse {
  data?: ListMCPOfficialCatalogData;
  code: number;
  msg: string;
}

export interface ListMCPToolRegistryEntriesRequest {
  space_id: string;
}

export interface GetMCPToolServerRequest {
  server_id: string;
}

export interface TestMCPToolCallRequest {
  server_id: string;
  tool_name: string;
  arguments: string;
}

export interface DiscoverMCPToolServerRequest {
  server_id: string;
}

export interface ExportMCPToolServerRequest {
  server_id: string;
}

export interface ListMCPToolAuditEventsRequest {
  server_id: string;
  limit?: number;
  cursor?: string;
}

export interface ListMCPToolServersData {
  servers: MCPToolServer[];
  total: number;
  can_manage: boolean;
}

export interface MCPToolRegistryEntry {
  name: string;
  source: string;
  category: string;
  visibility: string;
  server_id: string;
  server_name: string;
  tool_name: string;
  description: string;
  input_schema: string;
  enabled: boolean;
  health_status: string;
  health_checked_at: number;
  health_latency_ms: number;
  health_error: string;
}

export interface ListMCPToolRegistryEntriesData {
  tools: MCPToolRegistryEntry[];
  total: number;
}

export interface TestMCPToolCallData {
  status: string;
  output: string;
  latency_ms: number;
}

export interface DiscoverMCPToolServerData {
  tools: MCPToolDefinition[];
  resources: MCPResource[];
  prompts: MCPPrompt[];
}

export interface ExportMCPToolServerData {
  name: string;
  description: string;
  server_type: string;
  config: string;
  tools: MCPToolDefinition[];
  resources: MCPResource[];
  prompts: MCPPrompt[];
}

export interface MCPToolAuditEvent {
  event_id: string;
  actor_id: string;
  tool_name: string;
  status: 'pending' | 'success' | 'failed';
  latency_ms: number;
  error_code?: string;
  error_summary?: string;
  created_at: number;
  completed_at?: number;
}

export interface ListMCPToolAuditEventsData {
  events: MCPToolAuditEvent[];
  next_cursor: string;
}

export interface MCPToolServerResponse {
  data?: MCPToolServer;
  code: number;
  msg: string;
}

export interface ListMCPToolServersResponse {
  data?: ListMCPToolServersData;
  code: number;
  msg: string;
}

export interface ListMCPToolRegistryEntriesResponse {
  data?: ListMCPToolRegistryEntriesData;
  code: number;
  msg: string;
}

export interface TestMCPToolCallResponse {
  data?: TestMCPToolCallData;
  code: number;
  msg: string;
}

export interface DiscoverMCPToolServerResponse {
  data?: DiscoverMCPToolServerData;
  code: number;
  msg: string;
}

export interface ExportMCPToolServerResponse {
  data?: ExportMCPToolServerData;
  code: number;
  msg: string;
}

export interface ListMCPToolAuditEventsResponse {
  data?: ListMCPToolAuditEventsData;
  code: number;
  msg: string;
}

export const ListMCPToolServers = /*#__PURE__*/createAPI<
  ListMCPToolServersRequest,
  ListMCPToolServersResponse
>({
  url: '/api/workbench/mcp_tools',
  method: 'GET',
  name: 'ListMCPToolServers',
  reqType: 'ListMCPToolServersRequest',
  reqMapping: {
    query: ['space_id'],
  },
  resType: 'ListMCPToolServersResponse',
  schemaRoot: 'api://schemas/idl_workbench_tool',
  service: 'workbenchTool',
});

export const ListMCPOfficialCatalog = /*#__PURE__*/createAPI<
  ListMCPOfficialCatalogRequest,
  ListMCPOfficialCatalogResponse
>({
  url: '/api/workbench/mcp_tools/official_catalog',
  method: 'GET',
  name: 'ListMCPOfficialCatalog',
  reqType: 'ListMCPOfficialCatalogRequest',
  reqMapping: {
    query: ['space_id'],
  },
  resType: 'ListMCPOfficialCatalogResponse',
  schemaRoot: 'api://schemas/idl_workbench_tool',
  service: 'workbenchTool',
});

export const InstallMCPOfficialCatalog = /*#__PURE__*/createAPI<
  InstallMCPOfficialCatalogRequest,
  MCPToolServerResponse
>({
  url: '/api/workbench/mcp_tools/official_catalog/:catalog_id/install',
  method: 'POST',
  name: 'InstallMCPOfficialCatalog',
  reqType: 'InstallMCPOfficialCatalogRequest',
  reqMapping: {
    path: ['catalog_id'],
    body: ['space_id', 'credentials'],
  },
  resType: 'MCPToolServerResponse',
  schemaRoot: 'api://schemas/idl_workbench_tool',
  service: 'workbenchTool',
});

export const UpsertMCPToolServer = /*#__PURE__*/createAPI<
  UpsertMCPToolServerRequest,
  MCPToolServerResponse
>({
  url: '/api/workbench/mcp_tools',
  method: 'POST',
  name: 'UpsertMCPToolServer',
  reqType: 'UpsertMCPToolServerRequest',
  reqMapping: {
    body: [
      'server_id',
      'space_id',
      'name',
      'description',
      'server_type',
      'enabled',
      'config',
      'auth',
      'tools',
    ],
  },
  resType: 'MCPToolServerResponse',
  schemaRoot: 'api://schemas/idl_workbench_tool',
  service: 'workbenchTool',
});

export const ListMCPToolRegistryEntries = /*#__PURE__*/createAPI<
  ListMCPToolRegistryEntriesRequest,
  ListMCPToolRegistryEntriesResponse
>({
  url: '/api/workbench/mcp_tools/registry_entries',
  method: 'GET',
  name: 'ListMCPToolRegistryEntries',
  reqType: 'ListMCPToolRegistryEntriesRequest',
  reqMapping: {
    query: ['space_id'],
  },
  resType: 'ListMCPToolRegistryEntriesResponse',
  schemaRoot: 'api://schemas/idl_workbench_tool',
  service: 'workbenchTool',
});

export const GetMCPToolServer = /*#__PURE__*/createAPI<
  GetMCPToolServerRequest,
  MCPToolServerResponse
>({
  url: '/api/workbench/mcp_tools/:server_id',
  method: 'GET',
  name: 'GetMCPToolServer',
  reqType: 'GetMCPToolServerRequest',
  reqMapping: {
    path: ['server_id'],
  },
  resType: 'MCPToolServerResponse',
  schemaRoot: 'api://schemas/idl_workbench_tool',
  service: 'workbenchTool',
});

export const DeleteMCPToolServer = /*#__PURE__*/createAPI<
  GetMCPToolServerRequest,
  MCPToolServerResponse
>({
  url: '/api/workbench/mcp_tools/:server_id',
  method: 'DELETE',
  name: 'DeleteMCPToolServer',
  reqType: 'GetMCPToolServerRequest',
  reqMapping: {
    path: ['server_id'],
  },
  resType: 'MCPToolServerResponse',
  schemaRoot: 'api://schemas/idl_workbench_tool',
  service: 'workbenchTool',
});

export const TestMCPToolCall = /*#__PURE__*/createAPI<
  TestMCPToolCallRequest,
  TestMCPToolCallResponse
>({
  url: '/api/workbench/mcp_tools/:server_id/test_call',
  method: 'POST',
  name: 'TestMCPToolCall',
  reqType: 'TestMCPToolCallRequest',
  reqMapping: {
    path: ['server_id'],
    body: ['tool_name', 'arguments'],
  },
  resType: 'TestMCPToolCallResponse',
  schemaRoot: 'api://schemas/idl_workbench_tool',
  service: 'workbenchTool',
});

export const DiscoverMCPToolServer = /*#__PURE__*/createAPI<
  DiscoverMCPToolServerRequest,
  DiscoverMCPToolServerResponse
>({
  url: '/api/workbench/mcp_tools/:server_id/discover',
  method: 'POST',
  name: 'DiscoverMCPToolServer',
  reqType: 'DiscoverMCPToolServerRequest',
  reqMapping: {
    path: ['server_id'],
  },
  resType: 'DiscoverMCPToolServerResponse',
  schemaRoot: 'api://schemas/idl_workbench_tool',
  service: 'workbenchTool',
});

export const ExportMCPToolServer = /*#__PURE__*/createAPI<
  ExportMCPToolServerRequest,
  ExportMCPToolServerResponse
>({
  url: '/api/workbench/mcp_tools/:server_id/export',
  method: 'GET',
  name: 'ExportMCPToolServer',
  reqType: 'ExportMCPToolServerRequest',
  reqMapping: {
    path: ['server_id'],
  },
  resType: 'ExportMCPToolServerResponse',
  schemaRoot: 'api://schemas/idl_workbench_tool',
  service: 'workbenchTool',
});

export const ListMCPToolAuditEvents = /*#__PURE__*/createAPI<
  ListMCPToolAuditEventsRequest,
  ListMCPToolAuditEventsResponse
>({
  url: '/api/workbench/mcp_tools/:server_id/audit_events',
  method: 'GET',
  name: 'ListMCPToolAuditEvents',
  reqType: 'ListMCPToolAuditEventsRequest',
  reqMapping: {
    path: ['server_id'],
    query: ['limit', 'cursor'],
  },
  resType: 'ListMCPToolAuditEventsResponse',
  schemaRoot: 'api://schemas/idl_workbench_tool',
  service: 'workbenchTool',
});
